builder/layer.go

```diff
@@ -0,0 +1,272 @@
+package main
+
+import (
+	"bufio"
+	"crypto/sha512"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"os"
+	"path/filepath"
+	"syscall"
+	"time"
+
+	"github.com/docker/go-units"
+	"github.com/flynn/flynn/builder/store"
+	ct "github.com/flynn/flynn/controller/types"
+	"github.com/flynn/flynn/host/resource"
+	"github.com/flynn/flynn/host/types"
+	"github.com/flynn/flynn/pkg/cluster"
+	"github.com/flynn/flynn/pkg/exec"
+	"github.com/flynn/go-docopt"
+	"github.com/rminnich/go9p"
+	"gopkg.in/inconshreveable/log15.v2"
+)
+
+const (
+	LocalStoreDir = "/var/lib/flynn/local"
+
+	// use a larger default temp disk size as most build jobs need
+	// more than 100MB
+	DefaultTempDiskSize int64 = 1 * units.GiB
+)
+
+var cmdLayer = Command{
+	Run: runLayer,
+	Usage: `
+usage: flynn-build layer [options] <base> <script> [<inputs>...]
+
+Options:
+  --limits=<limits>   resource limits
+
+Build an image layer on top of <base> using <script>,
+serializing the layer JSON to STDOUT.
+
+A hash of the inputs is generated to determine if the layer has already been
+built (which prevents changes to the builder code rebuilding all images).
+
+Expects flynn-host to be running locally at 192.0.2.100 (which
+is started by host/start-build-daemon.sh).
+`[1:],
+}
+
+func runLayer(args *docopt.Args) error {
+	log := log15.New("component", "build")
+	log.SetHandler(log15.StreamHandler(os.Stderr, log15.LogfmtFormat()))
+
+	// use a local store
+	store, err := store.NewLocalStore(LocalStoreDir)
+	if err != nil {
+		log.Error("error initializing local store", "err", err)
+		return err
+	}
+
+	cwd, err := os.Getwd()
+	if err != nil {
+		log.Error("error getting current directory", "err", err)
+		return err
+	}
+
+	// check if we already created a layer for the given name + script + inputs
+	name := filepath.Base(cwd)
+	script := args.String["<script>"]
+	hash, err := generateHash(name, append([]string{script}, args.All["<inputs>"].([]string)...)...)
```

Comment:
I've added the base artifact but I'm not adding the builder artifact because then any changes in the builder code (e.g. changing `controller/types` which we do quite often in dev) leads to every image being rebuilt from scratch (this is the main reason I added the caching, otherwise tup would handle it).

Eventually we will use an out-of-tree builder (probably build using released flynn-host / flynn-build binaries), in which case we can use their hashes here.